### PR TITLE
Build iOS Rust code as an XCFramework.

### PR DESCRIPTION
**Pull request by rfk**
(Opening as a draft for early visibility)

This is an experiment in making our iOS build process more compatible
with M1 Mac build environments, and easier to consume via Swift Packages.

In our existing iOS build setup, we compile together all of the Rust
code and all of the wrapping Swift code into a single module named
"MozillaAppServices". This is a bit of a violation of clean package
separation, because it means that we risk name conflicts in the Swift
code between different components.

In this new setup, we don't pre-compile any of the Swift code. Instead
we build only the Rust code, package it along with its associated headers,
tell Swift to treat this as a module named "MozillaRustComponents".
Each individual Swift wrapper can then `import MozillaRustComponents`
and get access to the low-level FFI functions that it needs.

An additional benefit of this setup is that we can distribute it as
a `.xcframework` bundle for easy consumption on different platforms,
and we can build it without using XCode.

### CODE REVIEW

File: .circleci/config.yml
Comment (rfk) [inline .circleci/config.yml:391]: It's not clear to me that we actually want to build this on every single PR, especially if it's expensive to build. But it's useful to me for testing purposes to start out doing it that way, and also it's what we already do for the carthage archive.

**Comment by rfk** [comment]
OK! I seem to have managed to get CircleCI to build a `.zip` artifact using the new script and upload it:

* https://111888-129966583-gh.circle-artifacts.com/0/dist/MozillaRustComponents.xcframework.zip

Tomorrow, I will try pulling this in as a binary target from a Swift package and see if it'll work end-to-end.

**Comment by rfk** [comment]
> Tomorrow, I will try pulling this in as a binary target from a Swift package and see if it'll work end-to-end.

It works! The Swift Package repo over at https://github.com/mozilla/rust-components-swift/ is referencing this artifact from CircleCI and I can successfully build against it for a local simulator build.

The PR here is still draft because it's blocked by the UniFFI update in https://github.com/mozilla/application-services/pull/4377, but I believe it's ready for review in the meantime.

**Comment by rfk** [comment]
For testing whether this PR actually works, I recommend:

* Following the instructions in https://mozilla-hub.atlassian.net/browse/SYNC-2675 to test out a build with a published package, which as mentioned above is using an artifact produced by a CircleCI run on this PR.
* Follow the instructions in the README in this PR to try doing a local build and testing it in a consuming app, using [this fork of focus-ios](https://github.com/rfk/focus-ios/pull/2) as the consuming app.

I'm deliberately not saying any more detail about how to perform those tasks, because if there are parts of the procedure that aren't clear based on the docs in this PR and the rust-components-swift repo, then that's useful signal about things that need more docs :-)

File: .circleci/config.yml
Comment (tarikeshaq) [inline .circleci/config.yml:467]: Side comment: this is identical to the Carthage flow, which makes sense, I imagine we should probably briefly update https://mozilla.github.io/application-services/book/build-and-publish-pipeline.html?highlight=pipeline#application-services-build-and-publish-pipeline once those changes land (for the sake of time, we can get to it later)

File: .gitignore
Comment (tarikeshaq) [inline .gitignore:54]: 🙏

File: components/nimbus/ios/Nimbus/FeatureVariables.swift
Comment (tarikeshaq) [inline components/nimbus/ios/Nimbus/FeatureVariables.swift:6]: Just out of curiosity, I imagine something broke, it didn't compile and we needed to add this?

File: components/fxa-client/uniffi.toml
Comment (tarikeshaq) [inline components/fxa-client/uniffi.toml:7]: Not sure, but do we still want to keep this if we're not adding the `fxa_client` yet? (until we can get NSS built for M1)

File: megazords/ios-rust/README.md
Comment (tarikeshaq) [inline megazords/ios-rust/README.md:60]: Do we want to be explicit that the `#import` should follow the `#if canImport(...)` pattern? I think it's easy enough to figure out, especially if someone looks at one of the already existing components, but just thinking out loud

File: megazords/ios-rust/README.md
Comment (tarikeshaq) [inline megazords/ios-rust/README.md:83]: A super minor organizational thought (feel free to completely discarding this, I'm just making sure I note down everything I think of)

I read through this section and was first thinking "wait you can't do this locally without a PR?" until I hit this line, could be worth it to change the titles, maybe have this whole  section be a `## Testing` section, and have `### Testing Locally`, `### Testing a PR` and `### Testing a release` as separate sections - again for the sake of time, we can do this later ourselves 😄

File: megazords/ios-rust/build-xcframework.sh
Comment (tarikeshaq) [inline megazords/ios-rust/build-xcframework.sh:105]: 🤔 Could be a good spike for a later time (to explore running tests via `swift test` like you mentioned in our chat yesterday)

File: megazords/ios-rust/build-xcframework.sh
Comment (tarikeshaq) [inline megazords/ios-rust/build-xcframework.sh:136]: This could be a good ticket/spike on the Uniffi side for later 😄

File: megazords/ios-rust/build-xcframework.sh
Comment (tarikeshaq) [inline megazords/ios-rust/build-xcframework.sh:147]: This could be another follow up nice-to-have ticket!

**Review by tarikeshaq — COMMENTED** [review]
Took a look at the changes, looks good to me!

I'll spend some time getting this built locally and tested, then will come back and approve it!

Dropped a couple of thoughts, @rfk most of them are just reminders to myself to come back and create follow-up tickets (if they aren't already covered elsewhere)

File: components/fxa-client/uniffi.toml
Comment (tarikeshaq) [reply components/fxa-client/uniffi.toml:7]: Actually never mind, this more of a `uniffi` thing that it is a `Swift Packages` thing right?

File: megazords/ios-rust/README.md
Comment (tarikeshaq) [reply megazords/ios-rust/README.md:60]: Never mind! I have this the other way around 🤦 🤦

File: megazords/ios-rust/build-xcframework.sh
Comment (rfk) [reply megazords/ios-rust/build-xcframework.sh:105]: Right, exactly! I actually prototyped this a little bit and it seemed like it would work, but:

* Glean doesn't publish desktop builds, so we'd have to bring Glean into our megazord before we could do this
* It makes the compiled bundle even bigger, which may be a problem for production releases

File: megazords/ios-rust/README.md
Comment (rfk) [reply megazords/ios-rust/README.md:83]: Good idea, thanks.

File: components/fxa-client/uniffi.toml
Comment (rfk) [reply components/fxa-client/uniffi.toml:7]: It won't do any harm here if we leave it, but also it's only required because of how we do the XCFramework build, so it might make sense to take it out just to avoid confusion.

File: components/nimbus/ios/Nimbus/FeatureVariables.swift
Comment (rfk) [reply components/nimbus/ios/Nimbus/FeatureVariables.swift:6]: Ah yes, thanks for the reminder. The code in this file is using some classes from `UIKit` (I think something about image-handling) but wasn't importing it explicitly. That was fine in our old iOS build setup because something else in the megazord was import it, but it broke when trying to compile this as its own package.

**Review by skhamis — APPROVED** [review]
This is awesome! Really easy to understand and tested with both a "ci published" version as well as local ( built on an M1) and both worked flawlessly. Great foundation for an eventual switchover to full xcframeworks!

File: components/fxa-client/uniffi.toml
Comment (rfk) [reply components/fxa-client/uniffi.toml:7]: (I'm going to go ahead and remove it to avoid confusion)

**Comment by rfk** [comment]
Thanks both, I've updated to address a few of the docs-related comments.

**Pull request closed without merging**
